src/java.base/share/classes/jdk/internal/module/SystemModuleFinders.java

```diff
@@ -442,14 +450,24 @@ private InputStream toInputStream(ByteBuffer bb) { // ## -> ByteBuffer?
             }
         }
 
+        /**
+         * Returns the node for the given resource if found. If the name references
+         * a non-resource node, then {@code null} is returned.
+         */
+        private ImageReader.Node findResourceNode(ImageReader reader, String name) throws IOException {
+            Objects.requireNonNull(name);
+            if (closed) {
+                throw new IOException("ModuleReader is closed");
+            }
+            String nodeName = "/modules/" + module + "/" + name;
+            ImageReader.Node node = reader.findNode(nodeName);
+            return node != null && node.isResource() ? node : null;
+        }
+
         @Override
         public Optional<ByteBuffer> read(String name) throws IOException {
-            ImageLocation location = findImageLocation(name);
-            if (location != null) {
-                return Optional.of(SystemImage.reader().getResourceBuffer(location));
-            } else {
-                return Optional.empty();
-            }
+            ImageReader reader = SystemImage.reader();
+            return Optional.ofNullable(findResourceNode(reader, name)).map(reader::getResourceBuffer);
```

Comment:
Style wise, can you put the .map(...) on the next line so this is easier to read.